docs: x/distribution spec (#8928)
cosmos/cosmos-sdk: x/distribution/spec/04_messages.md

## MsgSetWithdrawAddress

> By default, the withdraw address is delegator address. If a delegator wants to change its withdraw address it must send `MsgSetWithdrawAddress`.

Reviewer: ```suggestion
By default, the withdraw address is the delegator address. To change its withdraw address, a delegator must send a `MsgSetWithdrawAddress` message.
```

> This is only possible if the parameter `WithdrawAddrEnabled` is set to `true`.

Reviewer: ```suggestion
Changing the withdraw address is possible only if the parameter `WithdrawAddrEnabled` is set to `true`.
```

> The withdraw address cannot be any of the module accounts. These are blocked from being withdraw addresses by being added to the distribution keeper's `blockedAddrs` array at initialization.

Reviewer: ```suggestion
The withdraw address cannot be any of the module accounts. These accounts are blocked from being withdraw addresses by being added to the distribution keeper's `blockedAddrs` array at initialization.
```

Response:

> +++ https://github.com/cosmos/cosmos-sdk/blob/v0.40.0/proto/cosmos/distribution/v1beta1/tx.proto#L29-L37

Reviewer 2: nit, I think we should update these to the latest version

Author: hmm, should we do that here? or would you want to script it so that it isn't done in a patchwork fashion?

## MsgWithdrawDelegatorReward

A delegator can withdraw its rewards.

> Internally in the distribution module, this is treated exactly as if the delegator simply started a new delegation of the same value, simultaneously removing the previous delegation with associated rewards.

Reviewer: ```suggestion
Internally in the distribution module, this transaction simultaneously removes the previous delegation with associated rewards, the same as if the delegator simply started a new delegation of the same value.
```

Reviewer: somehow putting what the transaction does makes more sense that saying what it's like

The rewards are sent immediately from the distribution `ModuleAccount` to the withdraw address.
Any remainder (truncated decimals) are sent to the community pool.
The starting height of the delegation is set to the current validator period, and the reference count for the previous period is decremented.
The amount withdrawn is deducted from the `ValidatorOutstandingRewards` variable for the validator.

In the F1 distribution, the total rewards are calculated per validator period, and a delegator receives a piece of those rewards in proportion to their stake in the validator.
In basic F1, the total rewards that all the delegators are entitled to between to periods is calculated the following way.

> Let `R(X)` be the total accumulated rewards up to period `X` divided by the tokens staked at that time, i.e. the reward ratio to multiply an individual delegators stake with to get their reward.

Reviewer 3: ```suggestion
Let `R(X)` be the total accumulated rewards up to period `X` divided by the tokens staked at that time. The delegator allocation is `R(X) * delegator_stake`.
```

Then the rewards for all the delegators for staking between periods `A` and `B` are `(R(B) - R(A)) * total stake`.

> However, this doesn't take slashing into account.

Reviewer: ```suggestion
However, these calculated rewards don't account for slashing.
```

Taking the slashes into account requires iteration.

> Let `F(X)` be the fraction a validator is to be slashed for the slashing event that happened at period `X`.
> If the validator was slashed at periods `P1, ..., PN`, where `A < P1`, `PN < B`, we calculate the individual delegator's rewards, `T(A, B)`, as follows:

Reviewer: ```suggestion
Let `F(X)` be the fraction a validator is to be slashed for a slashing event that happened at period `X`.
If the validator was slashed at periods `P1, ..., PN`, where `A < P1`, `PN < B`, the distribution module calculates the individual delegator's rewards, `T(A, B)`, as follows:
```

```
stake := initial stake
rewards := 0
previous := A
for P in P1, ..., PN`:
    rewards = (R(P) - previous) * stake
    stake = stake * F(P)
    previous = P
rewards = rewards + (R(B) - R(PN)) * stake
```

> Put another way, the historical rewards are calculated retroactively by playing back all the slashes, and attenuating the delegator's stake at each step.
> The final calculated stake should be equivalent to the actual staked coins in the delegation, within a margin of error due to rounding errors.

Reviewer: ```suggestion
The historical rewards are calculated retroactively by playing back all the slashes and then attenuating the delegator's stake at each step.
The final calculated stake is equivalent to the actual staked coins in the delegation with a margin of error due to rounding errors.
```

Response:

> +++ https://github.com/cosmos/cosmos-sdk/blob/v0.40.0/proto/cosmos/distribution/v1beta1/tx.proto#L42-L50

Reviewer 2: ditto

Author: @fedekunze See above

## WithdrawValidatorCommission

> The validator can send WithdrawValidatorCommission message to withdraw their accumulated commission.

Author: ```suggestion
The validator can send the WithdrawValidatorCommission message to withdraw their accumulated commission.
```

> The validator can send this message to withdraw their accumulated commission.

Reviewer 3: please use something specific here instead of `this`.

Reviewer: yes! add the message name and even a link to the message in the docs

> The commission is calculated every block during `BeginBlock`, so no iteration is required to withdraw.

Reviewer 3: ```suggestion
The commission is calculated in every block during `BeginBlock`, so no iteration is required to withdraw.
```

The amount withdrawn is deducted from the `ValidatorOutstandingRewards` variable for the validator.

> Only integer amounts can be sent, so if the accumulated awards have any decimals, the amount is truncated before it's sent, and the remainder is left to be withdrawn later.

Reviewer: ```suggestion
Only integer amounts can be sent. If the accumulated awards have decimals, the amount is truncated before the withdrawal is sent, and the remainder is left to be withdrawn later.
```

Reviewer: what is being sent? "the withdrawal" or the message for the withdraw tx?

Author: The withdrawal, i.e. the tokens are being sent to the delegator.

## FundCommunityPool

This message sends coins directly from the sender to the community pool.

> Expected to fail if for some reason the amount cannot be transferred from the sender to the distribution module account.

Reviewer: ```suggestion
The transaction fails if the amount cannot be transferred from the sender to the distribution module account.
```

```go
func (k Keeper) FundCommunityPool(ctx sdk.Context, amount sdk.Coins, sender sdk.AccAddress) error {
    if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, sender, types.ModuleName, amount); err != nil {
        return err
    }

	feePool := k.GetFeePool(ctx)
	feePool.CommunityPool = feePool.CommunityPool.Add(sdk.NewDecCoinsFromCoins(amount...)...)
	k.SetFeePool(ctx, feePool)

	return nil
}
```

> ## Common operations

Reviewer: ```suggestion
## Common distribution operations
```

> ## Common distribution operations

Reviewer: instead of stacking headings, can you introduce what common distribution operations are?

Author: ```suggestion
## Common distribution operations

These operations take place during many different messages.
```

### Initialize delegation

> Every time a delegation is changed, the rewards are withdrawn, and the delegation is reinitialized.
> Initializing a delegation means incrementing the validator period and keeping track of the starting period of the delegation.

Reviewer: ```suggestion
Each time a delegation is changed, the rewards are withdrawn and the delegation is reinitialized.
Initializing a delegation increments the validator period and keeps track of the starting period of the delegation.
```
